### DISCUSSION TITLE
how to add CASCADE deletion on schema level in prisma 2

### DISCUSSION THREAD

**Discussion by rajneesh3923** [prisma/prisma#2149] [Q&A] [answered]
how can i add CASCADE deletion on schema.prisma on a model in prisma 2., like we do in prisma 1

**Comment by dodas** [comment, marked as answer]
You can read about cascading deletes here:
https://www.prisma.io/docs/guides/database-workflows/cascading-deletes/postgresql

More specifically:

> Note: Deletion behaviors for relations are not yet supported in the Prisma schema so you don't see them anywhere. The behavior will still be enforced by the database though since that's where you configured it.

However, I found one issue:
Relation and therefore the foreign key field have to be Nullable in prisma schema.
Having them non-nullable while having `ON DELETE CASCADE` constraint in database leads to prisma throwing error:
```
The change you are trying to make would violate the required relation 'UserToPost' between the `User` and `Post` models.
```

**Reply by rajneesh3923** [reply]
According to your answer, I have made relation and the foreign key field nullable but when I delete a post the comments related to it sets the postId field on the comments table to null which is fine. But how can I set the CASCADE behavior on the database level?. so that the comments get deleted as well.

For your reference, I have generated the tables using prisma migrate command.

**Reply by dodas** [reply]
You need to set CONSTRAINT on database level. It is  described in the article I posted.
This is how I altered my table after it has been created with prisma migrate:
```sql
ALTER TABLE public.Post
DROP CONSTRAINT Post_user_fkey,
ADD CONSTRAINT Post_user_fkey
   FOREIGN KEY (user)
   REFERENCES public.User(user)
   ON DELETE CASCADE
   ON UPDATE CASCADE;
```

**Reply by rajneesh3923** [reply]
thank you very much

**Reply by techntrails** [reply]
> You can read about cascading deletes here:
> https://www.prisma.io/docs/guides/database-workflows/cascading-deletes/postgresql
> 
> More specifically:
> 
> > Note: Deletion behaviors for relations are not yet supported in the Prisma schema so you don't see them anywhere. The behavior will still be enforced by the database though since that's where you configured it.
> 
> However, I found one issue:
> Relation and therefore the foreign key field have to be Nullable in prisma schema.
> Having them non-nullable while having `ON DELETE CASCADE` constraint in database leads to prisma throwing error:
> 
> ```
> The change you are trying to make would violate the required relation 'UserToPost' between the `User` and `Post` models.
> ```

saved me a lot of time!

**Reply by electrovir** [reply]
> https://www.prisma.io/docs/guides/database-workflows/cascading-deletes/postgresql

This page no longer exists. I'm not sure what it covered, but I'm guessing the latest replacement is https://www.prisma.io/docs/orm/prisma-schema/data-model/relations/referential-actions

**Comment by spcfdoho** [comment]
@dodas I got the same issue and i have solved it. thanks. But if relation and foreign key should be nullable, we could create an orphaned post right?

**Comment by dodas** [comment]
@spcfdoho there is one open: https://github.com/prisma/prisma/issues/2057

**Comment by pranshushah** [comment]
from version 2.26  onDelete and onUpdate is the preview feature. more about [this](https://newreleases.io/project/github/prisma/prisma/release/2.26.0)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
